sql/analyzer/hoist_select_exists.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2022 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package analyzer
+
+import (
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+	"github.com/dolthub/go-mysql-server/sql/transform"
+)
+
+// hoistSelectExists merges a WHERE EXISTS subquery scope with its outer
+// scope when the subquery filters on columns from the outer scope.
+//
+// For example:
+// select * from a where exists (select 1 from b where a.x = b.x)
+// =>
+// select * from a semi join b on a.x = b.x
+func hoistSelectExists(
+	ctx *sql.Context,
+	a *Analyzer,
+	n sql.Node,
+	scope *Scope,
+	sel RuleSelector,
+) (sql.Node, transform.TreeIdentity, error) {
+	return transform.Node(n, func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		f, ok := n.(*plan.Filter)
+		if !ok {
+			return n, transform.SameTree, nil
+		}
+
+		left, right, outerFilters, joinType := pluckCorrelatedExistsSubquery(f)
+		if len(outerFilters) == 0 {
+			return n, transform.SameTree, nil
+		}
+		if right == nil || left == nil {
+			panic("unexpected empty scope")
+		}
+
+		if p, ok := right.(*plan.Project); ok {
+			right = p.Child
+		}
+
+		if gb, ok := right.(*plan.GroupBy); ok {
+			right = gb.Child
+		}
+
+		outerFilters, _, err := FixFieldIndexesOnExpressions(ctx, scope, a, append(left.Schema(), right.Schema()...), outerFilters...)
+		if err != nil {
+			return n, transform.SameTree, err
+		}
+
+		switch joinType {
+		case plan.AntiJoinType:
+			return plan.NewAntiJoin(left, right, expression.JoinAnd(outerFilters...)), transform.NewTree, nil
+		case plan.SemiJoinType:
+			return plan.NewSemiJoin(left, right, expression.JoinAnd(outerFilters...)), transform.NewTree, nil
+		default:
+			panic("expected SemiJoinType or AntiJoinType")
+		}
+	})
+}
+
+// pluckCorrelatedExistsSubquery scans a filter for [note] WHERE EXISTS, and then attempts to
```

Comment:
what is `[note]`

Reply:
autocorrect i think, meant `[NOT] WHERE EXISTS`